pkg/internal/ingress/ingress.go

```diff
@@ -0,0 +1,143 @@
+
+	"github.com/jetstack/cert-manager/pkg/controller"
+)
+
+// keep an internal cache of known API types so calls to the discovery API are kept to a minimum
+// https://pkg.go.dev/sync/atomic#example-Value-ReadMostly
+var (
+	knownAPIVersionCache atomic.Value
+	cacheLock            sync.Mutex
+)
+
+// InternalIngressCreateUpdater mimics a client-go networking/v1 or
+// networking/v1beta1 Interface.
+type InternalIngressCreateUpdater interface {
+	Ingresses(namespace string) InternalIngressInterface
+}
+
+// InternalIngressInterface mimics a client-go networking/v1/IngressInterface
+// It always returns a *networkingv1.Ingress, so when implementing this you must convert any type of
+// Ingress into a v1.Ingress.
+type InternalIngressInterface interface {
+	Create(ctx context.Context, ingress *networkingv1.Ingress, opts metav1.CreateOptions) (*networkingv1.Ingress, error)
+	Update(ctx context.Context, ingress *networkingv1.Ingress, opts metav1.UpdateOptions) (*networkingv1.Ingress, error)
+	Delete(ctx context.Context, name string, opts metav1.DeleteOptions) error
+}
+
+// InternalIngressLister mimics a client-go networking/v1/IngressLister.
+type InternalIngressLister interface {
+	List(selector labels.Selector) (ret []*networkingv1.Ingress, err error)
+	Ingresses(namespace string) InternalIngressNamespaceLister
+}
+
+// InternalIngressNamespaceLister mimics a client-go networking/v1/IngressNamespaceLister
+type InternalIngressNamespaceLister interface {
+	List(selector labels.Selector) (ret []*networkingv1.Ingress, err error)
+	Get(name string) (*networkingv1.Ingress, error)
+}
+
+// NewListerInformer returns an InternalIngressLister configured for v1 or v1beta1 ingresses depending on the
+// API Versions available in the discovery client.
+func NewListerInformer(ctx *controller.Context) (InternalIngressLister, cache.SharedIndexInformer, error) {
+	if hasVersion(ctx.DiscoveryClient, networkingv1.SchemeGroupVersion.String()) {
+		return &v1Lister{
+				lister: ctx.KubeSharedInformerFactory.Networking().V1().Ingresses().Lister(),
+			},
+			ctx.KubeSharedInformerFactory.Networking().V1().Ingresses().Informer(),
+			nil
+	} else if hasVersion(ctx.DiscoveryClient, networkingv1beta1.SchemeGroupVersion.String()) {
+		sch := runtime.NewScheme()
+		clientgoscheme.AddToScheme(sch)
+		return &v1beta1Lister{
+				scheme: sch,
+				lister: ctx.KubeSharedInformerFactory.Networking().V1beta1().Ingresses().Lister(),
+			},
+			ctx.KubeSharedInformerFactory.Networking().V1beta1().Ingresses().Informer(),
+			nil
+	}
+	return nil, nil, fmt.Errorf("neither %s or %s have any APIResources", networkingv1.SchemeGroupVersion, networkingv1beta1.SchemeGroupVersion)
+}
+
+// NewCreateUpdater returns an InternalIngressCreateUpdater configured for v1 or v1beta1 ingresses depending on the
+// versions available in the discovery client
+func NewCreateUpdater(ctx *controller.Context, discoveryOverrides ...discovery.DiscoveryInterface) (InternalIngressCreateUpdater, error) {
```

Comment:
suggestion: why do we return an interface here? The Go proverb "accept interfaces, return structs" would hold here, since `NewCreateUpdater` won't ever be used to initialize the struct itself.

Reply:
Can update that, I usually stick the interface in while developing so it doesn't compile if I miss an implementation.

Reply:
Oh, I return the interface because I return one of 2 structs depending on which ingress APIs are available.

Reply:
Aaah right, got it 😅